Sources_v3/StreamChatUI/MessageActionsPopup/ChatMessageActionsVC.swift

```diff
@@ -0,0 +1,174 @@
+    func chatMessageActionsVC(_ vc: ChatMessageActionsVC<ExtraData>, didTapOnThreadReplyFor message: _ChatMessage<ExtraData>)
+    func chatMessageActionsVC(_ vc: ChatMessageActionsVC<ExtraData>, didTapOnEdit message: _ChatMessage<ExtraData>)
+    func chatMessageActionsVCDidFinish(_ vc: ChatMessageActionsVC<ExtraData>)
+}
+
+open class ChatMessageActionsVC<ExtraData: ExtraDataTypes>: ViewController, UIConfigProvider {
+    public var messageController: _ChatMessageController<ExtraData>!
+    public var delegate: Delegate? // swiftlint:disable:this weak_delegate
+    public private(set) lazy var router = uiConfig.navigation.messageActionsRouter.init(rootViewController: self)
+
+    private var message: _ChatMessage<ExtraData>? {
+        messageController.message
+    }
+
+    // MARK: - Subviews
+
+    private lazy var messageActionView = uiConfig
+        .messageList
+        .messageActionsSubviews
+        .actionsView
+        .init()
+        .withoutAutoresizingMaskConstraints
+
+    // MARK: - Life Cycle
+
+    override open func setUpLayout() {
+        view.embed(messageActionView)
+    }
+
+    override open func updateContent() {
+        messageActionView.actionItems = messageActions
+    }
+
+    // MARK: - Actions
+
+    open var messageActions: [ChatMessageActionItem] {
+        guard
+            let currentUser = messageController.client.currentUserController().currentUser,
+            let message = message,
+            message.deletedAt == nil
+        else { return [] }
+
+        let editAction: ChatMessageActionItem = .edit { [weak self] in self?.handleEditAction() }
+        let deleteAction: ChatMessageActionItem = .delete { [weak self] in self?.handleDeleteAction() }
+
+        switch message.localState {
+        case nil:
+            var actions: [ChatMessageActionItem] = [
+                .inlineReply { [weak self] in self?.handleInlineReplyAction() },
+                .threadReply { [weak self] in self?.handleThreadReplyAction() },
+                .copy { [weak self] in self?.handleCopyAction() }
+            ]
+
+            if message.isSentByCurrentUser {
+                actions += [editAction, deleteAction]
+            } else if currentUser.mutedUsers.contains(message.author) {
+                actions.append(.unmuteUser { [weak self] in self?.handleUnmuteAuthorAction() })
+            } else {
+                actions.append(.muteUser { [weak self] in self?.handleMuteAuthorAction() })
+            }
+
+            return actions
+        case .pendingSend, .sendingFailed, .pendingSync, .syncingFailed, .deletingFailed:
+            return [
+                message.localState == .sendingFailed ? .resend { [weak self] in self?.handleResendAction() } : nil,
+                editAction,
+                deleteAction
+            ].compactMap { $0 }
+        case .sending, .syncing, .deleting:
+            return []
+        }
+    }
```

Comment:
Question: How crazy of an idea would it be to pass the `[ChatMessageActionItem]` from the outside including the execution closure - instead of using a delegate to propagate the action back?

Something like:
```swift
let messageActionsVC = ChatMessageActionsVC()
messageActionsVC.messageActions = [
  .editAction { .. },
  .deleteAction { .. },
  ...
] 
```
It'd mean having no delegate, no type-erasing, and a happier life 🎉

Reply:
Initially it was exactly like you described, we were passing `[ChatMessageActionItem]` to controller. In such implementation the `ChatMessageActionsVC` is dummy and does nothing expect forwarding actions to the view, but all the logic for creating set of available actions (depending on message state) and handlers is declared in `ChatMessageListVC`.

The obvious choice was to incapsulate the actions in some other type (actions provider/handler). But thinking about the type name I considered `ChatMessageActionsController` which is close to `ChatMessageActionsVC`. So I put this part of code back to `ChatMessageActionsVC` 😄

Reply:
@evsaev In that case, why don't `ChatMessageActionsVC` have its own MessageController where it can perform the actions? We can have as many MessageControllers as we need. I just don't see the point in using the delegate here.